### PR TITLE
[image-manipulator] multiple actions (web) and transparent background fill (web/android)

### PR DESCRIPTION
**Pull request by mnightingale**
# Why

There are a few issues I'd identified that were affecting my app using this package.

- [web] only the last action is returned [`manipulateWithActionAsync(uri || output!.uri, action, _options)`](https://github.com/expo/expo/blob/0edb5e5a76f300b55bbc29a5937b244872935b20/packages/expo-image-manipulator/src/ExpoImageManipulator.web.ts#L330)  
`uri` is the original  
`output!.uri` would be the output of the last action, but `uri` is always defined so it never gets used.
- [web] the second issue is linked to the first, it's a bit inefficient when multiple actions are used because each calls `toDataURL` (slow)  
I've resolved this by each action working on a new canvas and the action reducer passing the resulting canvas to the next action (similar to the Android approach). 
This refactor results in one call to `toDataURL` after all the actions have been processed and I think simplifies the code/logic a bit.
- [web][android] When you manipulate a PNG with transparency and save as JPEG you end up with a black background (iOS you get a white background).
I've resolved this by creating a `fill` action (implemented on all platforms) that is automatically appended to your actions (`actions.push({ fill: '#ffffff' });`) if you request a JPEG and the platform is not iOS.
- Web `console.warn` if you pass the `compress` argument for `PNG`, moved it to the validator (still a console.warn) - I don't think it's necessary though **(thoughts?)**

# Remaining issues (could be separate PR)

- [ ] returns { uri: base64, base64 } [docs](https://docs.expo.io/versions/latest/sdk/imagemanipulator/#returns) say the base64 should not include `data:image/xxx;base64,` (and shouldn't be present unless requested)

# How

- [Snack](https://snack.expo.dev/@mnightingale/expo-image-manipulator-transpareny-and-actions)
- Made a diff to use what the Snack demonstrates in bare-expo/ncl https://gist.github.com/mnightingale/0e4f0188fb4f123c8496b829cc674a49 replaces the existing api one because it doesnt seem to like having a api & component screen with the same name.

# Test Plan

- Added some tests for validator
- https://snack.expo.dev/@mnightingale/expo-image-manipulator-transpareny-and-actions

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by mnightingale** [comment]
@bbarthec before I update the docs and request a review what are your thoughts on the couple of issues I raised above?
- use of `console.warn` that was previously only shown on web, it seems unnecessary to me since the browser just doesn't use it and prior to the change in this PR ios/android were not warning about it.
- `uri` as base64 is inefficient (~40% bloated or similar) but I wouldn't change that unless every package was consistent (maybe like the base64 option have a `blob: true` or `outputFormat: 'blob'` etc. to avoid a breaking change)
But for now I'd just make the base64 property compliant with what the docs say (and only return it if requested).

File: packages/expo-image-manipulator/src/validators.ts [hunk 1]
```diff
@@ -107,6 +116,9 @@ export function validateSaveOptions({ base64, compress, format }: SaveOptions):
     if (compress < 0 || compress > 1) {
       throw new TypeError('The "compress" argument must be a number between 0 and 1');
     }
+    if (format === SaveFormat.PNG) {
+      console.warn('The "compress" argument is not supported with png format');
```
Comment (mnightingale) [inline packages/expo-image-manipulator/src/validators.ts:120]: Seems unnecessary

File: packages/expo-image-manipulator/src/ExpoImageManipulator.web.ts [hunk 2]
```diff
@@ -188,19 +188,10 @@ function getContext(canvas: HTMLCanvasElement): CanvasRenderingContext2D {
...
-    const quality = Math.min(1, Math.max(0, options.compress ?? 1));
-    base64 = canvas.toDataURL('image/' + format, quality);
-  } else {
-    // defaults to PNG with no loss
-    base64 = canvas.toDataURL();
-  }
+function getResults(canvas: HTMLCanvasElement, options: SaveOptions): ImageResult {
+  const { format = SaveFormat.JPEG, compress = 1 } = options;
+  const quality = Math.min(1, Math.max(0, compress));
+  const base64 = canvas.toDataURL('image/' + format, quality);
```
Comment (mnightingale) [inline packages/expo-image-manipulator/src/ExpoImageManipulator.web.ts:194]: base64 property (not uri) shouldn't include `data:image/xxx;base64,` and should only be present if requested

**Review by bbarthec — CHANGES_REQUESTED** [review]
Hey @mnightingale 👋 
I'm sorry that I've kept you waiting for review for this long 🙇
Overall the code looks good and PR is nicely shaped, but it's too complicated and touches too many issues simultaneously. I'd like you to split it in multiple PRs; each PR should focus on one consistent change to make it easier  to review and be sure about the introduced changes.
Please do that and I'll happily review and merge them 🙏

File: packages/expo-image-manipulator/src/validators.ts (hunk 1 above)
Comment (bbarthec) [reply packages/expo-image-manipulator/src/validators.ts:120]: If that's the unreachable state then we can confidently remove it 💪

File: packages/expo-image-manipulator/src/ExpoImageManipulator.web.ts (hunk 2 above)
Comment (bbarthec) [reply packages/expo-image-manipulator/src/ExpoImageManipulator.web.ts:194]: Docs might be missing some aspect and I think it should be compared with the mobile platforms' behaviour and follow their behaviour. Our major goal is to provide unified API across all supported platforms (in some cases that might be unavailable, but in this case we can ensure it works the same everywhere I believe).

**Pull request closed without merging**